### ISSUE TITLE
Recursive data structures cause crash when used in `var2str` or `rpc`

### ISSUE THREAD

**Issue by Jummit** [godotengine/godot#51071]
### Godot version

7f4eafa6f28e3aa3f36eb358ec3e0b69b1803240

### System information

Arch Linux

### Issue description

When passing a recursive Dictionary or Array to either `var2str` or `rpc`, Godot crashes with a segmentation fault.

This is similar to #48289

### Steps to reproduce

```gdscript
var d = []
d.append(d)
var2str(d) # crash
```

### Minimal reproduction project

[recursive_crash.zip](https://github.com/godotengine/godot/files/6908659/recursive_crash.zip)

**Comment by Buri** [comment]
This can also be seen with tree like structures: 
```
lass Foo:
	var bar
	func _init():
		bar = Bar.new(self)
class Bar:
	var foo
	func _init(f):
		foo = f

func _ready():
	var f = Foo.new()
	var2str(f) # crash
```

**Comment by akien-mga** [comment]
Fixed in 4.0-stable and later:
```
ERROR: Max recursion reached
   at: write (core/variant/variant_parser.cpp:1961)
```
